packages/redis/src/RedisStreamSink.ts

```diff
@@ -0,0 +1,96 @@
+import {
+    IOutputSink,
+    IPublishedMessage,
+    IOutputSinkGuarantees,
+    IRequireInitialization,
+    IDisposable,
+    IComponentContext,
+    makeLifecycle,
+    Lifecycle,
+    IMetrics,
+    failSpan,
+    DefaultComponentContext,
+    OutputSinkConsistencyLevel,
+} from "@walmartlabs/cookie-cutter-core";
+import { Span, Tags, Tracer } from "opentracing";
+
+import { IRedisInputStreamOptions, redisClient, IRedisClient, RedisMetadata } from ".";
+import { RedisOpenTracingTagKeys, RedisMetrics, RedisMetricResults } from "./RedisClient";
+
+export class RedisStreamSink
+    implements IOutputSink<IPublishedMessage>, IRequireInitialization, IDisposable {
+    public guarantees: IOutputSinkGuarantees;
+    private client: Lifecycle<IRedisClient>;
+    private tracer: Tracer;
+    private metrics: IMetrics;
+    private spanOperationName: string = "Redis Output Sink Client Call";
+
+    constructor(private readonly config: IRedisInputStreamOptions) {
+        this.tracer = DefaultComponentContext.tracer;
+        this.metrics = DefaultComponentContext.metrics;
+        this.guarantees = {
+            consistency: OutputSinkConsistencyLevel.None,
+            idempotent: false,
+        };
+    }
+    async sink(
+        output: IterableIterator<IPublishedMessage>,
+        bail: (err: any) => never
+    ): Promise<void> {
+        let span: Span;
+        try {
+            for (const msg of output) {
+                span = this.tracer.startSpan(this.spanOperationName, {
+                    childOf: msg.spanContext,
+                });
+
+                this.spanLogAndSetTags(span, this.config.db, this.config.writeStream);
+
+                await this.client.xAddObject(
+                    span,
+                    msg.message.type,
+                    this.config.writeStream,
+                    RedisMetadata.OutputSinkStreamKey,
+                    msg.message.payload
+                );
+
+                this.metrics.increment(RedisMetrics.XAdd, {
```

Comment:
i think you already increment this metric in the client itself ... this might be doube-incrementing it

Reply:
same comment for the span, it already creates the child span

Reply:
Which would place would you prefer for the span and increment to occur?

Reply:
i think in the client makes most sense